sklearn/_build_utils/cythonize.py

```diff
@@ -0,0 +1,192 @@
+
+Checks pyx files to see if they have been changed relative to their
+corresponding C files.  If they have, then runs cython on these files to
+recreate the C files.
+
+The script detects changes in the pyx files using checksums [or hashes] stored
+in a database file
+
+Simple script to invoke Cython on all .pyx
+files; while waiting for a proper build system. Uses file hashes to
+figure out if rebuild is needed.
+
+It is called by ./setup.py sdist so that sdist package can be installed without
+cython
+
+Originally written by Dag Sverre Seljebotn, and adapted from statsmodel 0.6.1
+(Modified BSD 3-clause)
+
+We copied it for scikit-learn.
+
+Note: this script does not check any of the dependent C libraries; it only
+operates on the Cython .pyx files.
+"""
+# author: Arthur Mensch
+# license: BSD
+
+from __future__ import division, print_function, absolute_import
+
+import os
+import re
+import sys
+import hashlib
+import subprocess
+
+HASH_FILE = 'cythonize.dat'
+DEFAULT_ROOT = 'sklearn'
+
+# WindowsError is not defined on unix systems
+try:
+    WindowsError
+except NameError:
+    WindowsError = None
+
+
+#
+# Rules
+#
+def process_pyx(fromfile, tofile):
+    try:
+        from Cython.Compiler.Version import version as cython_version
+        from distutils.version import LooseVersion
+        if LooseVersion(cython_version) < LooseVersion('0.21'):
+            raise Exception('Building scikit-learn requires Cython >= 0.21')
+
+    except ImportError:
+        pass
+
+    flags = ['--fast-fail']
+    if tofile.endswith('.cpp'):
+        flags += ['--cplus']
+
+    try:
+        try:
+            r = subprocess.call(['cython'] + flags + ["-o", tofile, fromfile])
+            if r != 0:
+                raise Exception('Cython failed')
+        except OSError:
+            # There are ways of installing Cython that don't result in a cython
+            # executable on the path, see gh-2397.
+            r = subprocess.call([sys.executable, '-c',
+                                 'import sys; from Cython.Compiler.Main '
+                                 'import setuptools_main as main;'
+                                 ' sys.exit(main())'] + flags +
+                                ["-o", tofile, fromfile])
+            if r != 0:
+                raise Exception('Cython failed')
+    except OSError:
+        raise OSError('Cython needs to be installed')
+
+
+rules = {
+    '.pyx': process_pyx,
+}
+
+
+#
+# Hash db
+#
+def load_hashes(filename):
+    # Return { filename : (sha1 of input, sha1 of output) }
+    if os.path.isfile(filename):
+        hashes = {}
+        with open(filename, 'r') as f:
+            for line in f:
+                filename, inhash, outhash = line.split()
+                hashes[filename] = (inhash, outhash)
+    else:
+        hashes = {}
+    return hashes
+
+
+def save_hashes(hash_db, filename):
+    with open(filename, 'w') as f:
+        for key, value in sorted(hash_db.items()):
```

Comment:
is it necessary that the `.dat` file is sorted? reading it as a dict will make it unsorted anyway.